In [2]:
# 24.09.10
import pandas as pd
import numpy as np

In [4]:
data = {
    "도시": ["서울", "서울", "서울", "부산", "부산", "부산", "인천", "인천"],
    "연도": ["2015", "2010", "2005", "2015", "2010", "2005", "2015", "2010"],
    "인구": [9904312, 9631482, 9762546, 3448737, 3393191, 3512547, 2890451, 263203],
    "지역": ["수도권", "수도권", "수도권", "경상권", "경상권", "경상권", "수도권", "수도권"]
}
columns = ["도시", "연도", "인구", "지역"]
df1 = pd.DataFrame(data, columns=columns)
df1

도시    연도       인구   지역
0  서울  2015  9904312  수도권
1  서울  2010  9631482  수도권
2  서울  2005  9762546  수도권
3  부산  2015  3448737  경상권
4  부산  2010  3393191  경상권
5  부산  2005  3512547  경상권
6  인천  2015  2890451  수도권
7  인천  2010   263203  수도권

In [7]:
# 피벗팅 pivot (자주 쓰임)
# 피봇테이블 : 데이터프레임 특정 열들을 행/열/데이터로 만드는 테이블
df1.pivot(index= '도시', columns= '연도', values= '인구')

연도       2005       2010       2015
도시                                 
부산  3512547.0  3393191.0  3448737.0
서울  9762546.0  9631482.0  9904312.0
인천        NaN   263203.0  2890451.0

In [11]:
# set_index
df1.set_index(['도시','연도'])[['인구']].unstack()

인구                      
연도       2005       2010       2015
도시                                 
부산  3512547.0  3393191.0  3448737.0
서울  9762546.0  9631482.0  9904312.0
인천        NaN   263203.0  2890451.0

In [16]:
# df1.pivot(index='지역', columns='연도', values='인구') # index가 중복되는 데이터를 보유함으로 에러 출력
df1.pivot(index=['도시','지역'], columns='연도', values='인구')
# 위와 같은 멀티 인덱스를 갖도록 수정
# 행/열 인덱스는 데이터를 검색하기 위한 기준 => 값이 1개만 존재해야 함

연도           2005       2010       2015
도시 지역                                  
부산 경상권  3512547.0  3393191.0  3448737.0
서울 수도권  9762546.0  9631482.0  9904312.0
인천 수도권        NaN   263203.0  2890451.0

In [18]:
np.random.seed(0)
df2 = pd.DataFrame({
    'key1': ['A', 'A', 'B', 'B', 'A'],
    'key2': ['one', 'two', 'one', 'two', 'one'],
    'data1': [1, 2, 3, 4, 5],
    'data2': [10, 20, 30, 40, 50]
})
df2

key1 key2  data1  data2
0    A  one      1     10
1    A  two      2     20
2    B  one      3     30
3    B  two      4     40
4    A  one      5     50

In [23]:
# groupby
mygroup = df2.groupby('key1')
mygroup.groups # groups 쓰는 이유는 안쓰면 주소값만 출력력

{'A': [0, 1, 4], 'B': [2, 3]}

In [25]:
mygroup.sum() # 문자열은 합산이 안되서 미출력됨

key2  data1  data2
key1                         
A     onetwoone      8     80
B        onetwo      7     70

In [27]:
df2

key1 key2  data1  data2
0    A  one      1     10
1    A  two      2     20
2    B  one      3     30
3    B  two      4     40
4    A  one      5     50

In [36]:
# df2의 key1을 기준으로 그룹화+data1의 합계
df2.groupby('key1')['data1'].sum()

df2.data1.groupby(df2.key1).sum() # 위와 동일한 출력
# df2.data1.groupby('key1').sum() # data1한테 key1이 없고 data1의 열을 이미 사용했기 때문에 오류

# df2.groupby('key1').sum().data1 # 권장하지 않는 방법

key1
A    8
B    7
Name: data1, dtype: int64

In [39]:
df2.groupby(['key1','key2'])['data1'].sum()

# 위와 같은 출력
# df2.groupby(['key1','key2']).data1.sum()

df2.data1.groupby([df2.key1, df2.key2]).sum()

key1  key2
A     one     6
      two     2
B     one     3
      two     4
Name: data1, dtype: int64

In [43]:
df2.groupby(['key1', 'key2']).sum().data1

key1  key2
A     one     6
      two     2
B     one     3
      two     4
Name: data1, dtype: int64

In [62]:
df2

key1 key2  data1  data2
0    A  one      1     10
1    A  two      2     20
2    B  one      3     30
3    B  two      4     40
4    A  one      5     50

In [76]:
df2.groupby(['key1','key2'])['data1'].sum().unstack()

key2  one  two
key1          
A       6    2
B       3    4

In [86]:
df1
df1.groupby(['지역','연도'])['인구'].sum().unstack()

연도      2005     2010      2015
지역                             
경상권  3512547  3393191   3448737
수도권  9762546  9894685  12794763

In [5]:
import seaborn as sns
iris = sns.load_dataset('iris')
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [90]:
iris.groupby('species').sum()

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa             250.3        171.4          73.1         12.3
versicolor         296.8        138.5         213.0         66.3
virginica          329.4        148.7         277.6        101.3

In [93]:
iris.groupby('species').mean() # 그룹별 각 열에 대한 평균값

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa             5.006        3.428         1.462        0.246
versicolor         5.936        2.770         4.260        1.326
virginica          6.588        2.974         5.552        2.026

In [96]:
iris.groupby('species').median() # 중위값

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa               5.0          3.4          1.50          0.2
versicolor           5.9          2.8          4.35          1.3
virginica            6.5          3.0          5.55          2.0

In [98]:
iris.groupby('species').min()

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa               4.3          2.3           1.0          0.1
versicolor           4.9          2.0           3.0          1.0
virginica            4.9          2.2           4.5          1.4

In [100]:
iris.groupby('species').max()

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa               5.8          4.4           1.9          0.6
versicolor           7.0          3.4           5.1          1.8
virginica            7.9          3.8           6.9          2.5

In [102]:
iris.groupby('species').std() # 표준편차

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa          0.352490     0.379064      0.173664     0.105386
versicolor      0.516171     0.313798      0.469911     0.197753
virginica       0.635880     0.322497      0.551895     0.274650

In [104]:
iris.groupby(iris.species).var() # 분산

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa          0.124249     0.143690      0.030159     0.011106
versicolor      0.266433     0.098469      0.220816     0.039106
virginica       0.404343     0.104004      0.304588     0.075433

In [107]:
iris.groupby('species').quantile() # 4분위수, 중위수와 동일하게 출력

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa               5.0          3.4          1.50          0.2
versicolor           5.9          2.8          4.35          1.3
virginica            6.5          3.0          5.55          2.0

In [110]:
iris.groupby(iris.species).median() # quantile과 동일함

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa               5.0          3.4          1.50          0.2
versicolor           5.9          2.8          4.35          1.3
virginica            6.5          3.0          5.55          2.0

In [115]:
iris.groupby('species').first() # 그룹의 첫번째 데이터를 이용
iris.groupby('species').last() # 그룹의 마지막 데이터를 이용
# 잘 안쓰임

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa               5.0          3.3           1.4          0.2
versicolor           5.7          2.8           4.1          1.3
virginica            5.9          3.0           5.1          1.8

In [117]:
iris.groupby('species').groups # 종류별로 데이터 50개

{'setosa': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49], 'versicolor': [50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99], 'virginica': [100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149]}

In [119]:
iris.groupby('species').count() # 셀 별로 데이터 건수 확인

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa                50           50            50           50
versicolor            50           50            50           50
virginica             50           50            50           50

In [121]:
iris.groupby('species').size() # 그룹별 데이터 갯수

# count, size 차이
# - count는 열별로, size는 그룹별 개수 출력
# - size는 NaN을 포함

species
setosa        50
versicolor    50
virginica     50
dtype: int64

In [123]:
titanic = sns.load_dataset('titanic')
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [125]:
titanic.info()
titanic.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [128]:
titanic.age.count()

714

In [132]:
titanic.groupby('survived').age.count() # NaN 미포함
titanic.groupby('survived').age.size()

survived
0    549
1    342
Name: age, dtype: int64

In [142]:
#
iris.groupby('species').max() - iris.groupby('species').min()

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa               1.5          2.1           0.9          0.5
versicolor           2.1          1.4           2.1          0.8
virginica            3.0          1.6           2.4          1.1

In [146]:
iris.groupby('species').describe()

sepal_length                                              \
                  count   mean       std  min    25%  50%  75%  max   
species                                                               
setosa             50.0  5.006  0.352490  4.3  4.800  5.0  5.2  5.8   
versicolor         50.0  5.936  0.516171  4.9  5.600  5.9  6.3  7.0   
virginica          50.0  6.588  0.635880  4.9  6.225  6.5  6.9  7.9   

           sepal_width         ... petal_length      petal_width         \
                 count   mean  ...          75%  max       count   mean   
species                        ...                                        
setosa            50.0  3.428  ...        1.575  1.9        50.0  0.246   
versicolor        50.0  2.770  ...        4.600  5.1        50.0  1.326   
virginica         50.0  2.974  ...        5.875  6.9        50.0  2.026   

                                               
                 std  min  25%  50%  75%  max  
species                                        
setosa      0.105386  0.1  0.2  0.2  0.3  0.6  
versicolor  0.197753  1.0  1.2  1.3  1.5  1.8  
virginica   0.274650  1.4  1.8  2.0  2.3  2.5  

[3 rows x 32 columns]

@@@@@@@@@@@@@@@@@

# 숙제
# 종별로 그룹화를 한 다음, 꽃의 길이(petal_length)가 가장 큰 3개의 데이터 추출(총 9개 추출)
		        sepal_length	sepal_width     petal_length    petal_width	species
        species						
setosa	    4	            4.8	        3.4	              1.9	    0.2	    setosa
            4	            5.1	        3.8	              1.9	    0.4	    setosa
            3	            5.1	        3.3	              1.7	    0.5	    setosa
versicolor  83	            6.0	        2.7	              5.1	    1.6	    versicolor
            77	            6.7	        3.0	              5.0	    1.7	    versicolor
            72	            6.3	        2.5	              4.9	    1.5	    versicolor
virginica   118	            7.7	        2.6	              6.9	    2.3	    virginica
            117	            7.7	        3.8	              6.7	    2.2	    virginica
            122	            7.7	        2.8	              6.7	    2.0	    virginica

In [ ]:
import seaborn as sns
iris = sns.load_dataset('iris')
iris.head()

In [17]:
# 종별로 그룹화 한 다음, 꽃의 길이(petal_length)가 큰 데이터 3개 추출(총 9개)
iris.groupby('species').max()

# iris.apply(maxData)

sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa               5.8          4.4           1.9          0.6
versicolor           7.0          3.4           5.1          1.8
virginica            7.9          3.8           6.9          2.5

@@@@@@@@@@@@@@@@@@

In [164]:
df1

도시    연도       인구   지역
0  서울  2015  9904312  수도권
1  서울  2010  9631482  수도권
2  서울  2005  9762546  수도권
3  부산  2015  3448737  경상권
4  부산  2010  3393191  경상권
5  부산  2005  3512547  경상권
6  인천  2015  2890451  수도권
7  인천  2010   263203  수도권

In [166]:
df1.pivot_table('인구','도시','연도')
# df1.pivot(index='도시', columns='연도', values='인구') 와 같음

연도       2005       2010       2015
도시                                 
부산  3512547.0  3393191.0  3448737.0
서울  9762546.0  9631482.0  9904312.0
인천        NaN   263203.0  2890451.0

In [172]:
df1.pivot_table('인구','도시','연도',margins=True) # 집계함수 진행 -> 평균 구하기
# 3.451492e+06 => 3.451492 10**6
# pd.options.display.float_format = '{:.2f}'.fortmat 하면 소수점 2자리까지만 출력

연도        2005       2010       2015           All
도시                                                
부산   3512547.0  3393191.0  3448737.0  3.451492e+06
서울   9762546.0  9631482.0  9904312.0  9.766113e+06
인천         NaN   263203.0  2890451.0  1.576827e+06
All  6637546.5  4429292.0  5414500.0  5.350809e+06

In [173]:
df1.pivot_table('인구','도시','연도',margins=True, aggfunc='max', fill_value='1') # 최대값 추출
# aggfunc과 fill_value는 직접 다른 함수를 선언하여 입력하거나 다른 함수를 입력해도 무관

C:\Users\wag19\AppData\Local\Temp\ipykernel_17032\2814800887.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df1.pivot_table('인구','도시','연도',margins=True, aggfunc='max', fill_value='1') # 최대값 추출


연도      2005     2010     2015      All
도시                                     
부산   3512547  3393191  3448737  3512547
서울   9762546  9631482  9904312  9904312
인천         1   263203  2890451  2890451
All  9762546  9631482  9904312  9904312

In [174]:
df1.pivot_table('인구',['연도','도시'])

인구
연도   도시           
2005 부산  3512547.0
     서울  9762546.0
2010 부산  3393191.0
     서울  9631482.0
     인천   263203.0
2015 부산  3448737.0
     서울  9904312.0
     인천  2890451.0